CpHeader command failure

With the most recent commit (ef95e55f4732e26c83b4a8cb266f21c6897b309c) my build fails almost immediately in a `CpHeader` step with this log:

```
Create product structure
/bin/mkdir -p /Users/ksmiley/Library/Developer/Xcode/DerivedData/Lyft-dvyneupirgclsnggqtgohnrdrmxl/Build/Products/Debug-iphoneos/Alamofire-iOS/Alamofire.framework/Headers

CpHeader /Users/ksmiley/Library/Developer/Xcode/DerivedData/Lyft-dvyneupirgclsnggqtgohnrdrmxl/Build/Products/Debug-iphoneos/Alamofire-iOS/Alamofire.framework/Headers/Alamofire-iOS-umbrella.h '/Users/ksmiley/dev/phoenix/Pods/Alamofire/../Target Support Files/Alamofire-iOS/Alamofire-iOS-umbrella.h'
    cd /Users/ksmiley/dev/phoenix/Pods
    builtin-copy -exclude CVS -exclude .svn -exclude .git -exclude .hg /Users/ksmiley/dev/phoenix/Pods/Alamofire/../Target Support Files/Alamofire-iOS/Alamofire-iOS-umbrella.h /Users/ksmiley/Library/Developer/Xcode/DerivedData/Lyft-dvyneupirgclsnggqtgohnrdrmxl/Build/Products/Debug-iphoneos/Alamofire-iOS/Alamofire.framework/Headers

** BUILD FAILED **

The following build commands failed:
    CpHeader /Users/ksmiley/Library/Developer/Xcode/DerivedData/Lyft-dvyneupirgclsnggqtgohnrdrmxl/Build/Products/Debug-iphoneos/Alamofire-iOS/Alamofire.framework/Headers/Alamofire-iOS-umbrella.h '/Users/ksmiley/dev/phoenix/Pods/Alamofire/../Target Support Files/Alamofire-iOS/Alamofire-iOS-umbrella.h'
(1 failure)
```

I did a quick bisect and determined this is the first commit with this issue. Let me know if I can provide any more info about this!

--- a/Libraries/xcdriver/Sources/BuildAction.cpp
+++ b/Libraries/xcdriver/Sources/BuildAction.cpp
@@ -13,6 +13,7 @@
 #include <xcexecution/NinjaExecutor.h>
 #include <xcexecution/SimpleExecutor.h>
 #include <xcformatter/DefaultFormatter.h>
+#include <xcformatter/NullFormatter.h>
 #include <builtin/Registry.h>
 #include <libutil/Base.h>
 #include <libutil/Filesystem.h>
@@ -42,6 +43,9 @@ CreateFormatter(ext::optional<std::string> const &formatter)
         bool color = isatty(fileno(stdout));
 
         auto formatter = xcformatter::DefaultFormatter::Create(color);
+        return std::static_pointer_cast<xcformatter::Formatter>(formatter);
+    } else if (*formatter == "null") {
+        auto formatter = xcformatter::NullFormatter::Create();
         return std::static_pointer_cast<xcformatter::Formatter>(formatter);
     }
 

--- a/Libraries/xcexecution/CMakeLists.txt
+++ b/Libraries/xcexecution/CMakeLists.txt
@@ -18,3 +18,7 @@ add_library(xcexecution SHARED
 target_link_libraries(xcexecution PUBLIC xcformatter pbxbuild xcscheme xcworkspace pbxproj pbxsetting process util dependency ninja builtin)
 target_include_directories(xcexecution PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/Headers")
 install(TARGETS xcexecution DESTINATION usr/lib)
+
+if (BUILD_TESTING)
+  ADD_UNIT_GTEST(xcexecution SimpleExecutor Tests/test_SimpleExecutor.cpp)
+endif ()

--- a/Libraries/xcexecution/Headers/xcexecution/SimpleExecutor.h
+++ b/Libraries/xcexecution/Headers/xcexecution/SimpleExecutor.h
@@ -35,7 +35,7 @@ public:
         pbxbuild::Build::Environment const &buildEnvironment,
         Parameters const &buildParameters);
 
-private:
+public:
     bool writeAuxiliaryFiles(
         libutil::Filesystem *filesystem,
         pbxproj::PBX::Target::shared_ptr const &target,
@@ -45,8 +45,7 @@ private:
         process::Context const *processContext,
         process::Launcher *processLauncher,
         libutil::Filesystem *filesystem,
-        pbxproj::PBX::Target::shared_ptr const &target,
-        pbxbuild::Target::Environment const &targetEnvironment,
+        std::vector<std::string> const &executablePaths,
         std::vector<pbxbuild::Tool::Invocation> const &orderedInvocations,
         bool createProductStructure);
     std::pair<bool, std::vector<pbxbuild::Tool::Invocation>> buildTarget(

--- a/Libraries/xcexecution/Sources/SimpleExecutor.cpp
+++ b/Libraries/xcexecution/Sources/SimpleExecutor.cpp
@@ -217,8 +217,7 @@ performInvocations(
     process::Context const *processContext,
     process::Launcher *processLauncher,
     Filesystem *filesystem,
-    pbxproj::PBX::Target::shared_ptr const &target,
-    pbxbuild::Target::Environment const &targetEnvironment,
+    std::vector<std::string> const &executablePaths,
     std::vector<pbxbuild::Tool::Invocation> const &orderedInvocations,
     bool createProductStructure)
 {
@@ -258,7 +257,8 @@ performInvocations(
                         processContext->groupID(),
                         processContext->userName(),
                         processContext->groupName());
-                    success = driver->run(&context, filesystem) != 0;
+                    int exitCode = driver->run(&context, filesystem);
+                    success = (exitCode == 0);
 
                     xcformatter::Formatter::Print(_formatter->finishInvocation(invocation, *builtin, createProductStructure));
                 } else {
@@ -273,7 +273,7 @@ performInvocations(
                         path = external;
                     }
                 } else {
-                    path = filesystem->findExecutable(*external, targetEnvironment.executablePaths());
+                    path = filesystem->findExecutable(*external, executablePaths);
                 }
 
                 if (path) {
@@ -329,13 +329,13 @@ buildTarget(
     }
 
     xcformatter::Formatter::Print(_formatter->beginCreateProductStructure(target));
-    std::pair<bool, std::vector<pbxbuild::Tool::Invocation>> structureResult = performInvocations(processContext, processLauncher, filesystem, target, targetEnvironment, *orderedInvocations, true);
+    std::pair<bool, std::vector<pbxbuild::Tool::Invocation>> structureResult = performInvocations(processContext, processLauncher, filesystem, targetEnvironment.executablePaths(), *orderedInvocations, true);
     xcformatter::Formatter::Print(_formatter->finishCreateProductStructure(target));
     if (!structureResult.first) {
         return structureResult;
     }
 
-    std::pair<bool, std::vector<pbxbuild::Tool::Invocation>> invocationsResult = performInvocations(processContext, processLauncher, filesystem, target, targetEnvironment, *orderedInvocations, false);
+    std::pair<bool, std::vector<pbxbuild::Tool::Invocation>> invocationsResult = performInvocations(processContext, processLauncher, filesystem, targetEnvironment.executablePaths(), *orderedInvocations, false);
     if (!invocationsResult.first) {
         return invocationsResult;
     }

--- a/Libraries/xcexecution/Tests/test_SimpleExecutor.cpp
+++ b/Libraries/xcexecution/Tests/test_SimpleExecutor.cpp
@@ -1,0 +1,145 @@
+/**
+ Copyright (c) 2015-present, Facebook, Inc.
+ All rights reserved.
+
+ This source code is licensed under the BSD-style license found in the
+ LICENSE file in the root directory of this source tree. An additional grant
+ of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#include <gtest/gtest.h>
+#include <xcexecution/SimpleExecutor.h>
+#include <xcformatter/NullFormatter.h>
+#include <pbxbuild/Tool/Invocation.h>
+#include <builtin/Driver.h>
+#include <builtin/Registry.h>
+#include <process/MemoryContext.h>
+#include <process/MemoryLauncher.h>
+#include <libutil/MemoryFilesystem.h>
+
+using xcexecution::SimpleExecutor;
+using libutil::Filesystem;
+using libutil::MemoryFilesystem;
+
+class Driver : public builtin::Driver {
+public:
+    using Impl = std::function<int(process::Context const *processContext, Filesystem *filesystem)>;
+
+private:
+    std::string _name;
+    Impl        _impl;
+
+public:
+    Driver(std::string const &name, Impl const &impl) :
+        _name(name),
+        _impl(impl)
+    {
+    }
+
+public:
+    virtual std::string name()
+    { return _name; }
+
+public:
+    virtual int run(process::Context const *processContext, Filesystem *filesystem)
+    { return _impl(processContext, filesystem); }
+};
+
+TEST(SimpleExecutor, PropagateToolResult)
+{
+    /* Create in-memory execution environment. */
+    auto filesystem = MemoryFilesystem({
+        MemoryFilesystem::Entry::File("fail-tool", std::vector<uint8_t>()),
+        MemoryFilesystem::Entry::File("success-tool", std::vector<uint8_t>()),
+    });
+
+    auto launcher = process::MemoryLauncher({
+        { "/fail-tool", [](Filesystem *filesystem, process::Context const *context) -> ext::optional<int> {
+            return 1;
+        } },
+        { "/success-tool", [](Filesystem *filesystem, process::Context const *context) -> ext::optional<int> {
+            return 0;
+        } },
+    });
+
+    auto registry = builtin::Registry::Create({
+        std::static_pointer_cast<builtin::Driver>(std::make_shared<Driver>("builtin-fail", [](process::Context const *context, Filesystem *filesystem) -> int {
+            return 1;
+        })),
+        std::static_pointer_cast<builtin::Driver>(std::make_shared<Driver>("builtin-success", [](process::Context const *context, Filesystem *filesystem) -> int {
+            return 0;
+        })),
+    });
+
+    auto context = process::MemoryContext(
+        "",
+        "/",
+        std::vector<std::string>(),
+        std::unordered_map<std::string, std::string>(),
+        0,
+        0,
+        "user",
+        "group");
+
+    /* Create invocations to execute. */
+    auto builtinSuccess = pbxbuild::Tool::Invocation();
+    builtinSuccess.executable() = pbxbuild::Tool::Invocation::Executable::Builtin("builtin-success");
+    auto builtinFail = pbxbuild::Tool::Invocation();
+    builtinFail.executable() = pbxbuild::Tool::Invocation::Executable::Builtin("builtin-fail");
+
+    auto externalSuccess = pbxbuild::Tool::Invocation();
+    externalSuccess.executable() = pbxbuild::Tool::Invocation::Executable::External("success-tool");
+    auto externalFail = pbxbuild::Tool::Invocation();
+    externalFail.executable() = pbxbuild::Tool::Invocation::Executable::External("fail-tool");
+
+    /* Create test executor. */
+    auto formatter = xcformatter::NullFormatter::Create();
+    std::vector<std::string> const executablePaths = { "/" };
+    SimpleExecutor executor = SimpleExecutor(formatter, false, registry);
+
+    /* Succeed if all tools succeed. */
+    auto success = executor.performInvocations(
+        &context,
+        &launcher,
+        &filesystem,
+        executablePaths,
+        {
+            builtinSuccess,
+            externalSuccess,
+        },
+        false);
+    ASSERT_TRUE(success.first);
+    EXPECT_EQ(success.second.size(), 0);
+
+    /* Fail if a tool fails. */
+    auto fail1 = executor.performInvocations(
+        &context,
+        &launcher,
+        &filesystem,
+        executablePaths,
+        {
+            externalFail,
+            builtinSuccess,
+            externalSuccess,
+        },
+        false);
+    ASSERT_FALSE(fail1.first);
+    EXPECT_EQ(fail1.second.size(), 1);
+
+    /* Fail if a tool fails, even if not first. */
+    auto fail2 = executor.performInvocations(
+        &context,
+        &launcher,
+        &filesystem,
+        executablePaths,
+        {
+            builtinSuccess,
+            builtinFail,
+            externalSuccess,
+            externalFail,
+        },
+        false);
+    ASSERT_FALSE(fail2.first);
+    EXPECT_EQ(fail2.second.size(), 1);
+}
+

--- a/Libraries/xcformatter/CMakeLists.txt
+++ b/Libraries/xcformatter/CMakeLists.txt
@@ -10,6 +10,7 @@
 add_library(xcformatter SHARED
             Sources/Formatter.cpp
             Sources/DefaultFormatter.cpp
+            Sources/NullFormatter.cpp
             )
 
 target_link_libraries(xcformatter PUBLIC pbxbuild pbxproj pbxsetting)

--- a/Libraries/xcformatter/Headers/xcformatter/NullFormatter.h
+++ b/Libraries/xcformatter/Headers/xcformatter/NullFormatter.h
@@ -1,0 +1,59 @@
+/**
+ Copyright (c) 2015-present, Facebook, Inc.
+ All rights reserved.
+
+ This source code is licensed under the BSD-style license found in the
+ LICENSE file in the root directory of this source tree. An additional grant
+ of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#ifndef __xcformatter_NullFormatter_h
+#define __xcformatter_NullFormatter_h
+
+#include <xcformatter/Formatter.h>
+
+namespace xcformatter {
+
+/*
+ * Formatter that prints no output.
+ */
+class NullFormatter : public Formatter {
+public:
+    NullFormatter();
+    virtual ~NullFormatter();
+
+public:
+    virtual std::string begin(pbxbuild::Build::Context const &buildContext);
+    virtual std::string success(pbxbuild::Build::Context const &buildContext);
+    virtual std::string failure(pbxbuild::Build::Context const &buildContext, std::vector<pbxbuild::Tool::Invocation> const &failingInvocations);
+
+public:
+    virtual std::string beginTarget(pbxbuild::Build::Context const &buildContext, pbxproj::PBX::Target::shared_ptr const &target);
+    virtual std::string finishTarget(pbxbuild::Build::Context const &buildContext, pbxproj::PBX::Target::shared_ptr const &target);
+
+public:
+    virtual std::string beginCheckDependencies(pbxproj::PBX::Target::shared_ptr const &target);
+    virtual std::string finishCheckDependencies(pbxproj::PBX::Target::shared_ptr const &target);
+
+public:
+    virtual std::string beginWriteAuxiliaryFiles(pbxproj::PBX::Target::shared_ptr const &target);
+    virtual std::string createAuxiliaryDirectory(std::string const &directory);
+    virtual std::string writeAuxiliaryFile(std::string const &file);
+    virtual std::string setAuxiliaryExecutable(std::string const &file);
+    virtual std::string finishWriteAuxiliaryFiles(pbxproj::PBX::Target::shared_ptr const &target);
+
+public:
+    virtual std::string beginCreateProductStructure(pbxproj::PBX::Target::shared_ptr const &target);
+    virtual std::string finishCreateProductStructure(pbxproj::PBX::Target::shared_ptr const &target);
+
+public:
+    virtual std::string beginInvocation(pbxbuild::Tool::Invocation const &invocation, std::string const &executable, bool simple);
+    virtual std::string finishInvocation(pbxbuild::Tool::Invocation const &invocation, std::string const &executable, bool simple);
+
+public:
+    static std::shared_ptr<NullFormatter> Create();
+};
+
+}
+
+#endif // !__xcformatter_NullFormatter_h

--- a/Libraries/xcformatter/Sources/NullFormatter.cpp
+++ b/Libraries/xcformatter/Sources/NullFormatter.cpp
@@ -1,0 +1,130 @@
+/**
+ Copyright (c) 2015-present, Facebook, Inc.
+ All rights reserved.
+
+ This source code is licensed under the BSD-style license found in the
+ LICENSE file in the root directory of this source tree. An additional grant
+ of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#include <xcformatter/NullFormatter.h>
+#include <pbxbuild/Tool/Invocation.h>
+#include <pbxbuild/Build/Context.h>
+
+#include <algorithm>
+
+using xcformatter::NullFormatter;
+
+NullFormatter::
+NullFormatter() :
+    Formatter()
+{
+}
+
+NullFormatter::
+~NullFormatter()
+{
+}
+
+std::string NullFormatter::
+begin(pbxbuild::Build::Context const &buildContext)
+{
+    return std::string();
+}
+
+std::string NullFormatter::
+success(pbxbuild::Build::Context const &buildContext)
+{
+    return std::string();
+}
+
+std::string NullFormatter::
+failure(pbxbuild::Build::Context const &buildContext, std::vector<pbxbuild::Tool::Invocation> const &failingInvocations)
+{
+    return std::string();
+}
+
+std::string NullFormatter::
+beginTarget(pbxbuild::Build::Context const &buildContext, pbxproj::PBX::Target::shared_ptr const &target)
+{
+    return std::string();
+}
+
+std::string NullFormatter::
+finishTarget(pbxbuild::Build::Context const &buildContext, pbxproj::PBX::Target::shared_ptr const &target)
+{
+    return std::string();
+}
+
+std::string NullFormatter::
+beginCheckDependencies(pbxproj::PBX::Target::shared_ptr const &target)
+{
+    return std::string();
+}
+
+std::string NullFormatter::
+finishCheckDependencies(pbxproj::PBX::Target::shared_ptr const &target)
+{
+    return std::string();
+}
+
+std::string NullFormatter::
+beginWriteAuxiliaryFiles(pbxproj::PBX::Target::shared_ptr const &target)
+{
+    return std::string();
+}
+
+std::string NullFormatter::
+createAuxiliaryDirectory(std::string const &directory)
+{
+    return std::string();
+}
+
+std::string NullFormatter::
+writeAuxiliaryFile(std::string const &file)
+{
+    return std::string();
+}
+
+std::string NullFormatter::
+setAuxiliaryExecutable(std::string const &file)
+{
+    return std::string();
+}
+
+std::string NullFormatter::
+finishWriteAuxiliaryFiles(pbxproj::PBX::Target::shared_ptr const &target)
+{
+    return std::string();
+}
+
+std::string NullFormatter::
+beginCreateProductStructure(pbxproj::PBX::Target::shared_ptr const &target)
+{
+    return std::string();
+}
+
+std::string NullFormatter::
+finishCreateProductStructure(pbxproj::PBX::Target::shared_ptr const &target)
+{
+    return std::string();
+}
+
+std::string NullFormatter::
+beginInvocation(pbxbuild::Tool::Invocation const &invocation, std::string const &executable, bool simple)
+{
+    return std::string();
+}
+
+std::string NullFormatter::
+finishInvocation(pbxbuild::Tool::Invocation const &invocation, std::string const &executable, bool simple)
+{
+    return std::string();
+}
+
+std::shared_ptr<NullFormatter> NullFormatter::
+Create()
+{
+    return std::make_shared<NullFormatter>();
+}
+

